optimization: Speed up Base58 encoding/decoding by 400%/200% via preliminary byte packing

l0rinc opened the pull request:
To mitigate the quadratic complexity inherent in the sequential byte division of the Base58 encoding process, this update aggregates characters into groups of 7 and 9, fitting them into 64-bit long integers.
This refinement utilizes the inherent efficiency of native division and modulo operations on 64-bit architectures, significantly optimizing computational overhead.
The benchmarks indicate a ~4.4x speedup for `Base58Encode`:

> make && ./src/bench/bench_bitcoin --filter=Base58Encode --min-time=1000

before:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               61.00 |       16,393,083.32 |    0.2% |      1.07 | `Base58Encode`

after:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               13.79 |       72,502,616.20 |    0.1% |      1.10 | `Base58Encode`

------

The same was applied to decoding, resulting in a ~2x speedup:

> make && ./src/bench/bench_bitcoin --filter=Base58Decode --min-time=1000

before:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               17.41 |       57,429,723.99 |    0.1% |      1.10 | `Base58Decode`

after:
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                8.80 |      113,590,754.53 |    0.1% |      1.11 | `Base58Decode`

-----

This also speed up [listunspent](https://github.com/bitcoin/bitcoin/pull/29473#issuecomment-2200977622) by >50%.

-----

See https://github.com/bitcoin/bitcoin/pull/29473#issuecomment-2200977622 for additional benchmarks and https://github.com/bitcoin/bitcoin/pull/30035 for additional tests.
I've also tested with with the following temp test (comparing the exact previous impl with the new one for random inputs: https://gist.github.com/paplorinc/96d8e355f6fef10ac79f62d89a6d9f19#file-base58_tests-cpp-L211-L249)

l0rinc committed:
Remove redundant zero-skipping logic in Base58 encode/decode

l0rinc committed:
Encode by iterating forward and dividing in-place

Instead of collecting the values to and empty b58, we're cloning the input without leading zeros and dividing it in-place, while doing the quadratic division from the most significant, non-zero digit, forward.

make && ./src/bench/bench_bitcoin --filter=Base58Encode --min-time=1000
Before:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               59.34 |       16,851,250.88 |    0.2% |      1.10 | `Base58Encode`
|               59.20 |       16,892,479.61 |    0.2% |      1.10 | `Base58Encode`
|               58.97 |       16,958,226.76 |    0.2% |      1.10 | `Base58Encode`
```
After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               40.80 |       24,508,402.06 |    0.1% |      1.10 | `Base58Encode`
|               40.83 |       24,489,762.49 |    0.2% |      1.10 | `Base58Encode`
|               39.71 |       25,182,310.62 |    0.2% |      1.10 | `Base58Encode`
```

and

make && ./src/bench/bench_bitcoin --filter=Base58CheckEncode --min-time=1000
Before:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               79.93 |       12,511,576.75 |    0.4% |      1.10 | `Base58CheckEncode`
|               79.49 |       12,580,136.21 |    0.1% |      1.10 | `Base58CheckEncode`
|               79.65 |       12,554,785.16 |    0.1% |      1.10 | `Base58CheckEncode`
```
After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               55.27 |       18,094,561.35 |    0.1% |      1.10 | `Base58CheckEncode`
|               55.41 |       18,046,778.32 |    0.1% |      1.10 | `Base58CheckEncode`
|               55.32 |       18,075,763.16 |    0.1% |      1.10 | `Base58CheckEncode`
```

Co-authored-by: optout21 <13562139+optout21@users.noreply.github.com>

l0rinc committed:
Speed up Base58 encoding with 64-bit packing

To mitigate the quadratic complexity inherent in the sequential byte division of the Base58 encoding process, this update aggregates characters into groups of 7, fitting them into 64-bit long integers.
This refinement utilizes the inherent efficiency of native division and modulo operations on 64-bit architectures, significantly optimizing computational overhead.
The benchmarks indicate a 4x speedup for `Base58CheckEncode` and `Base58Encode`:

make && ./src/bench/bench_bitcoin --filter=Base58Encode --min-time=1000
After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               14.32 |       69,831,031.44 |    0.1% |      1.10 | `Base58Encode`
|               14.32 |       69,811,995.18 |    0.1% |      1.10 | `Base58Encode`
|               14.35 |       69,662,527.88 |    0.5% |      1.10 | `Base58Encode`
```

make && ./src/bench/bench_bitcoin --filter=Base58CheckEncode --min-time=1000
After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               19.15 |       52,231,968.11 |    0.1% |      1.06 | `Base58CheckEncode`
|               19.13 |       52,269,345.54 |    0.4% |      1.05 | `Base58CheckEncode`
|               19.14 |       52,244,117.61 |    0.3% |      1.06 | `Base58CheckEncode`
```

paplorinc commented on src/base58.cpp:
```diff
@@ -93,8 +97,7 @@ std::string EncodeBase58(Span<const unsigned char> input)
         zeroes++;
     }
-    // Allocate enough space in big-endian base58 representation.
-    int size = input.size() * 138 / 100 + 1; // log(256) / log(58), rounded up.
```
`ceil(log(256)/log(58) * 100) = 137`, not 138.

paplorinc commented on src/base58.cpp:
```diff
@@ -48,8 +53,7 @@ static const int8_t mapBase58[256] = {
         psz++;
     }
-    // Allocate enough space in big-endian base256 representation.
-    int size = strlen(psz) * 733 /1000 + 1; // log(58) / log(256), rounded up.
```
`ceil(log(58)/log(256) * 1000)` is indeed `733`

paplorinc commented on src/base58.cpp:
```diff
@@ -35,6 +36,10 @@ static const int8_t mapBase58[256] = {
     -1,-1,-1,-1,-1,-1,-1,-1, -1,-1,-1,-1,-1,-1,-1,-1,
 };
 
+static constexpr int baseScale{1000};
```
we could use 1024 here for fast division - not sure it's worth the change

paplorinc commented on src/base58.cpp:
```diff
@@ -78,7 +78,6 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
         psz++;
     if (*psz != 0)
         return false;
-    // Skip leading zeroes in b256.
```
probably left over from a previous refactoring, there's no skipping here anymore

paplorinc commented on src/base58.cpp:
```diff
@@ -114,12 +113,9 @@ std::string EncodeBase58(Span<const unsigned char> input)
     }
-    // Skip leading zeroes in base58 result.
-    std::vector<unsigned char>::iterator it = b58.begin() + (size - length);
-    while (it != b58.end() && *it == 0)
```
we're collecting the zeros in the beginning, this was always false

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
-        assert(carry == 0);
-        length = i;
-        input = input.subspan(1);
+    auto inputClone = Clone(input, leadingZeros);
```
slightly slower but a lot simpler - preparing the scene for the main commit

paplorinc commented on src/base58.cpp:
```diff
@@ -88,34 +90,42 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
 
-std::vector<unsigned char> Clone(const Span<const unsigned char>& input, size_t leadingZeros) {
-    return std::vector<unsigned char>(input.begin() + leadingZeros, input.end());
+auto ConvertToLongArray(const Span<const unsigned char>& input, size_t leadingZeros, size_t groupSize) -> std::vector<int64_t> {
```
converting from `base(256)` to `base(256^7)` speeds up processing since we can do actual division and modulo on the resulting int64

paplorinc commented on src/base58.cpp:
```diff
@@ -18,8 +18,8 @@
 /** All alphanumeric characters except for "0", "I", "O", and "l" */
-static const char* pszBase58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
-static const int8_t mapBase58[256] = {
+static const char* PSZ_BASE58 = "123456789ABCDEFGHJKLMNPQRSTUVWXYZabcdefghijkmnopqrstuvwxyz";
```
unify the constants in a separate commit to leave the campground cleaner than I found it

paplorinc commented on src/test/data/base58_encode_decode.json:
```diff
@@ -12,5 +12,9 @@
+["000102030405060708090a0b0c0d0e0f101112131415161718191a1b1c1d1e1f202122232425262728292a2b2c2d2e2f303132333435363738393a3b3c3d3e3f404142434445464748494a4b4c4d4e4f505152535455565758595a5b5c5d5e5f606162636465666768696a6b6c6d6e6f707172737475767778797a7b7c7d7e7f808182838485868788898a8b8c8d8e8f909192939495969798999a9b9c9d9e9fa0a1a2a3a4a5a6a7a8a9aaabacadaeafb0b1b2b3b4b5b6b7b8b9babbbcbdbebfc0c1c2c3c4c5c ...
+["271F359E", "zzzzy"],
+["271F359F", "zzzzz"],
+["271F35A0", "211111"],
```
271F35A0 is 58^5 - and -2, -1, and +1 to make sure boundaries are handled correctly

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongArray(const Span<const unsigned char>& input, size_t leadingZeros, size_t GROUP_SIZE) -> std::vector<int64_t> {
```
convert to base56 quickly before converting to base58 to minimize the inner quadratic loop

Empact replied:
Looks like you are unintentionally aliasing `GROUP_SIZE` as both a constant and this argument.

paplorinc replied:
Hah, indeed, las minute refactoring, thanks!

l0rinc committed:
Unify size estimations in Base58 encoding/decoding

Note that the constants for size allocations have slightly changed after recalculating them

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongArray(const Span<const unsigned char>& input, size_t leadingZeros) -> std::vector<int64_t> {
```
convert to base56 quickly before converting to base58 to minimize the inner quadratic loop

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
+auto ConvertToLongArray(const Span<const unsigned char>& input, size_t leadingZeros) -> std::vector<int64_t> {
+    auto effectiveLength = input.size() - leadingZeros;
+    std::vector<int64_t> inputAsLongs((effectiveLength + (GROUP_SIZE - 1)) / GROUP_SIZE);
+    auto groupOffset = ((GROUP_SIZE - (effectiveLength % GROUP_SIZE)) % GROUP_SIZE) - leadingZeros;
```
Is there a floorMod in C++?
```
int FloorMod(int x, int y) {
    int r = x % y;
    return (x ^ y) < 0 && r != 0 ? r + y : r;
}
// ...
auto groupOffset = FloorMod(-effectiveLength, GROUP_SIZE) - leadingZeros;
```

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongs(const Span<const unsigned char>& input, size_t leadingZeros) -> std::vector<int64_t> {
```
converting from base(256) to base(256^7) speeds up processing since we can do actual division and modulo on the resulting int64

optout21 commented on src/base58.cpp:
```diff
@@ -84,44 +90,45 @@ static const int8_t mapBase58[256] = {
+    result.assign(leadingZeros, '1');
+
+    auto inputAsLongs = ConvertToLongs(input, leadingZeros);
+    for (auto i{0U}; i < inputAsLongs.size();) {
```
This construct with the two nested loops is not evident. Maybe some comment would be useful. The inner loop does not depend on the value of `i`. Is it repeated for as long as all `inputAsLongs` elements are reduced to 0?

paplorinc replied:
Fantastic comment, the internal loop should skip the leading zeros, indeed.

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
+    auto inputAsLongs = ConvertToLongs(input, leadingZeros);
+    for (auto i{0U}; i < inputAsLongs.size();) {
+        int64_t remainder{0};
+        for (auto j{i}; j < inputAsLongs.size(); ++j) { // Calculate next digit, dividing inputAsLongs
```
start division and next digit calculation from the biggest non-zero chunk

paplorinc commented on src/base58.cpp:
```diff
@@ -88,33 +90,43 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
         result += pszBase58[remainder];
-        while (i < inputClone.size() && inputClone[i] == 0) {
-            ++i;
+        while (i < inputAsLongs.size() && inputAsLongs[i] == 0) {
```
this speeds up processing more than I though

Empact commented:
I have some questions about this change, but to start I would encourage you to make the change a minimal one. I.e. avoid strictly unnecessary changes, such as those in the last commit. Merge the variable name changes with the commit they are introduced, and don't change other lines in the file for conventions.

The goal, again, is to ease review and avoid cause for controversy. Leaving the code unchanged is a good default if not significantly justified.

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
     return true;
 }
 
+auto Clone(const Span<const unsigned char>& input, int start) -> std::vector<unsigned char> {
```
extracted to a method to minimize diff with the next commit where we'll start grouping bytes

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
-        length = i;
-        input = input.subspan(1);
+    auto inputClone = Clone(input, leading);
+    for (auto i{0U}; i < inputClone.size();) {
```
iterating forward instead

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
+    auto inputClone = Clone(input, leading);
+    for (auto i{0U}; i < inputClone.size();) {
+        int64_t remainder{0};
+        for (auto j{i}; j < inputClone.size(); ++j) { // Calculate next digit, dividing inputAsLongs
```
dividing the cloned values in-place

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
+            remainder = accumulator % 58;
+        }
+        result += pszBase58[remainder];
+        while (i < inputClone.size() && inputClone[i] == 0) {
```
continuing the quadratic division from the highest available digit

paplorinc commented on src/base58.cpp:
```diff
@@ -87,40 +88,37 @@ static const int log256_58Ratio = ceil((std::log(256) / std::log(58)) * baseScal
-    while (it != b58.end())
-        str += pszBase58[*(it++)];
-    return str;
+    std::reverse(result.begin() + leading, result.end());
```
we're collecting directly into the result, but need a linear reversal at the end

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
     return true;
 }
 
+auto ConvertToLongs(const Span<const unsigned char>& input, int start) -> std::vector<int64_t> {
```
pack to 64 bit  quickly (enabling native division for the groups), before converting to base58 to minimize the inner quadratic loop

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
+auto ConvertToLongs(const Span<const unsigned char>& input, int start) -> std::vector<int64_t> {
+    int effectiveLength = input.size() - start;
+    std::vector<int64_t> inputAsLongs((effectiveLength + (encodeGroupSize - 1)) / encodeGroupSize);
+    int groupOffset = ((encodeGroupSize - (effectiveLength % encodeGroupSize)) % encodeGroupSize) - start;
```
Is there a `floorMod` in C++?
```
int FloorMod(int x, int y) {
    int r = x % y;
    return (x ^ y) < 0 && r != 0 ? r + y : r;
}
// ...
auto groupOffset = FloorMod(-effectiveLength, encodeGroupSize) - start;
```

fanquake commented:
Pleas stop leaving endless self review comments on PRs. It's not only unecessary, but everytime you comment here, it sends 4000 people an email. Some commentary / responding to review comment is obviously fine, but what you are currently doing is just turning the thread into a spammy/confusing mess, and I'd assume, making things potentially harder for reviewers.

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
 
+auto ConvertToLongs(const Span<const unsigned char>& input, int start) -> std::vector<int64_t> {
+    int effectiveLength = input.size() - start;
+    std::vector<int64_t> inputAsLongs((effectiveLength + (encodeGroupSize - 1)) / encodeGroupSize);
```
this is basically a ceil integer division, i.e. `(int) ceil(effectiveLength / (double) encodeGroupSize)`

paplorinc commented on src/base58.cpp:
```diff
@@ -84,44 +90,47 @@ static const int8_t mapBase58[256] = {
+    for (auto i{0U}; i < inputAsLongs.size();) {
+        int64_t remainder{0};
+        for (auto j{i}; j < inputAsLongs.size(); ++j) { // Calculate next digit, dividing inputAsLongs
+            auto accumulator = (remainder << (encodeGroupSize * 8)) | inputAsLongs[j];
```
we're using a `|` instead of a `+` since we're shifting by powers of two, so we're adding to zeros

Empact commented:
@paplorinc agree with @fanquake, a better place to articulate your motivations and reasoning is in the commit messages, that way they'll be retained for future reference via `git blame`, etc.

l0rinc committed:
Decode by iterating forward and dividing in-place

> make && ./src/bench/bench_bitcoin --filter=Base58Decode --min-time=1000
Before:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               17.50 |       57,141,622.65 |    0.1% |      1.10 | `Base58Decode`
|               17.42 |       57,392,223.96 |    0.0% |      1.10 | `Base58Decode`
|               17.43 |       57,358,655.44 |    0.2% |      1.10 | `Base58Decode`
```
After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|               16.37 |       61,093,842.90 |    0.2% |      1.10 | `Base58Decode`
|               16.37 |       61,100,514.64 |    0.1% |      1.10 | `Base58Decode`
|               16.38 |       61,046,275.93 |    0.1% |      1.10 | `Base58Decode`
```

l0rinc committed:
Speed up Base58 decoding with 64-bit packing

> make && ./src/bench/bench_bitcoin --filter=Base58Decode --min-time=1000

After:
```
|             ns/byte |              byte/s |    err% |     total | benchmark
|--------------------:|--------------------:|--------:|----------:|:----------
|                8.79 |      113,767,685.06 |    0.1% |      1.10 | `Base58Decode`
|                8.78 |      113,831,528.33 |    0.0% |      1.10 | `Base58Decode`
|                8.80 |      113,607,470.35 |    0.2% |      1.10 | `Base58Decode`
```

paplorinc commented on src/test/data/base58_encode_decode.json:
```diff
@@ -11,6 +11,13 @@
 ["00000000000000000000", "1111111111"],
+["00000000000000000000000000000000000000000000000000000000000000000000000000000000", "1111111111111111111111111111111111111111"],
+["00000000000000000000000000000000000000000000000000000000000000000000000000000001", "1111111111111111111111111111111111111112"],
+["0000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000000ec39d04c37e71e5d591881f6", "111111111111111111111111111111111111111111111111111111111111111111111111111111111111115TYzLYH1udmLdzCLM"],
```
I've added the random test cases that failed during development to make sure they're covered deterministically.
Also added ones which are testing the boundaries of powers of 58, since those cases happen less frequently in the randomized tests.

paplorinc replied:
Since some of these failed during development of this feature, I think it adds and extra layer of safety to see that they're passing after the change as well.
But if you think it's better to merge it before this change, I can split it out of course.

paplorinc commented:
@Sjors, this is the optimization I've mentioned to you in person, I'd appreciate a review if you think it's worthwhile:
<img src="https://github.com/bitcoin/bitcoin/assets/1841944/e6c76863-eb97-41d2-b518-14eccac4442a">

paplorinc commented:
> I don't think this is used in a critical path, to warrant a speedup?

It's literally the example from the [benchmarking docs](https://github.com/bitcoin/bitcoin/blob/master/doc/benchmarking.md): 
<img src="https://github.com/bitcoin/bitcoin/assets/1841944/822e347b-0064-4343-8292-bf753530a7e9">

which was specifically optimized in various other commits before: https://github.com/bitcoin/bitcoin/commit/bcab47bc1b2bfdd29f8b89f8a211755299938aea, https://github.com/bitcoin/bitcoin/commit/159ed95f7452ab5454e9d660f6f095b8476b1eed, https://github.com/bitcoin/bitcoin/commit/3252208cb10be645bae415c90fb2ed8217838490

paplorinc commented:
What's the point of the benchmarks in that case exactly?
Anyway, I can make the code slightly simpler and similarly performant, would that be welcome?

paplorinc commented:
Thanks @sipa, appreciate your comment!
How do I find small changes that are needed, welcome, not so controversial, and don't need 3 years of up-front studying?
That's why I like optimizations, since the desired behavior is already in the code and test and benchmarks and history, and I can have visible progress while learning about the project.

andrewtoth commented:
> How do I find small changes that are needed, welcome, not so controversial, and don't need 3 years of up-front studying?

@paplorinc I know you are past your first PR and issue, but most of the issues with this label would fit that criteria https://github.com/bitcoin/bitcoin/labels/good%20first%20issue.
Also, browsing open issues I'm sure you could find others that fulfill most of the criteria as well.

paplorinc commented:
@andrewtoth, that was the first thing I checked, couldn't find any reasonable "good first issue" - or other issue that piqued my interest. The problem has to be personal, otherwise the solution won't be honest.

@maflcko, you already suggested that multiple times, but my question wasn't how to review PRs.

I'll check if this PR improved `listunspent`, as suggested, if not, I'll just close this.

andrewtoth commented:
@paplorinc https://github.com/bitcoin/bitcoin/pull/28945 was abandoned and I don't think @fjahr has taken it over. Perhaps that would be interesting to you.

andrewtoth commented:
Ahh sorry thanks.

paplorinc commented:
> Ok, the reason for https://github.com/bitcoin/bitcoin/pull/7656#issue-139438690 was an improvement in listunspent. Seems fine, if this is still the case. But this will need to be checked first.

I've created 10k blocks in regtest and measured the performance of `listunspent` before and after:

```bash
killall bitcoind 2>/dev/null || true
./src/bitcoind -regtest -daemon
./src/bitcoin-cli -regtest createwallet "test_wallet"
./src/bitcoin-cli -regtest generatetoaddress 10000 $(./src/bitcoin-cli -regtest getnewaddress) >/dev/null

hyperfine \
  --warmup 3 --runs 10 \
  --parameter-list BRANCH master,paplorinc/optimize-base58-conversion \
  --prepare '
    killall bitcoind 2>/dev/null || true
    git checkout "{BRANCH}" && git reset --hard
    make -j10 >/dev/null
    ./src/bitcoind -regtest -daemon
    sleep 1
    ./src/bitcoin-cli -regtest loadwallet "test_wallet"
  ' \
  './src/bitcoin-cli -regtest listunspent'
```
which resulted in: 
```sql
Benchmark 1: ./src/bitcoin-cli -regtest listunspent (BRANCH = master)
  Time (mean ± σ):     378.7 ms ±   2.8 ms    [User: 71.7 ms, System: 13.9 ms]
  Range (min … max):   376.4 ms … 386.0 ms    10 runs
 
Benchmark 2: ./src/bitcoin-cli -regtest listunspent (BRANCH = paplorinc/optimize-base58-conversion)
  Time (mean ± σ):     244.0 ms ±   1.3 ms    [User: 71.4 ms, System: 13.7 ms]
  Range (min … max):   242.2 ms … 246.1 ms    10 runs
 
Summary
  ./src/bitcoin-cli -regtest listunspent (BRANCH = paplorinc/optimize-base58-conversion) ran
    1.55 ± 0.01 times faster than ./src/bitcoin-cli -regtest listunspent (BRANCH = master)
```

i.e. 55% faster `listunspent` for the given usecase.

paplorinc commented:
I was following the benchmarks to decide what to work on, but since you've mentioned this is an important usecase, I measured it as well.
Is there anything else you'd like me to measure or change in this pull request?

paplorinc commented:
@luke-jr, would this optimization be more welcome in https://github.com/bitcoin/libbase58 instead?

luke-jr commented:
>would this optimization be more welcome in https://github.com/bitcoin/libbase58 instead?

The algorithm used in libbase58 is different, so not sure this even applies. Kinda doubt it would be worth the time to port/review either, unless a library consumer cares about performance or the other criteria explained already here.

If you're just looking for things to do, extending libbase58 to Bech32 might be worth doing. See https://github.com/bitcoin/libbase58/issues/6

achow101 commented:
As this is basically a complete rewrite of the base58 encoding and decoding code, I don't think the review effort required to review this is worth it relative to the unimportance of these functions.

l0rinc commented:
I just found this was already attempted in a very similar way by @martinus a few years ago in https://github.com/bitcoin/bitcoin/pull/21176 - so now we have two versions, if this will turn out to be important again.
Closing, thanks for the reviews!
